File: server/models/ProfitLoss.js
const mongoose = require('mongoose');

const profitLossSchema = new mongoose.Schema(
  {
    period: { type: String, enum: ['benchmark', 'current', 'prior'], required: true },
    sales: {
      openingStock: { type: Number, default: 0 },
      purchases: { type: Number, default: 0 },
      wages: { type: Number, default: 0 },
      other: { type: Number, default: 0 },
      revenue: { type: Number, default: 0 },
    },
    cogs: {
      closingStock: { type: Number, default: 0 },
      totalCOGS: { type: Number, default: 0 },
      grossProfit: { type: Number, default: 0 },
    },
    expenses: { type: Object, default: {} },
    netProfit: { type: Number, default: 0 },
  },
  { timestamps: true }
);

module.exports = mongoose.model('ProfitLoss', profitLossSchema);

File: python-api/calculators/profit_loss.py
def calculate_profit_loss(data: dict) -> dict:
    revenue = float(data.get("revenue", 0))
    opening_stock = float(data.get("openingStock", 0))
    purchases = float(data.get("purchases", 0))
    closing_stock = float(data.get("closingStock", 0))
    total_expenses = float(data.get("totalExpenses", 0))

    cogs = opening_stock + purchases - closing_stock
    gross_profit = revenue - cogs
    net_profit = gross_profit - total_expenses
    gross_margin = (gross_profit / revenue * 100) if revenue else 0
    net_margin = (net_profit / revenue * 100) if revenue else 0

    return {
        "cogs": cogs,
        "grossProfit": gross_profit,
        "netProfit": net_profit,
        "grossMarginPercent": gross_margin,
        "netMarginPercent": net_margin,
    }
